### PR TITLE
REP-5274 Compare full index specs rather than just a subset.

### PR DESCRIPTION
**Pull request by FGasper**
Previously migration-verifier only checked the parts of index specs that go in a mongo.IndexSpecification. That struct is incomplete, though; it doesn’t, for example, include partial index filters. Thus, migration-verifier didn’t compare those parts of an index specification.

This changeset alters the index-checking logic to check full index specifications. Note that this comparison is rather “coarse” and may flag some specifications as mismatching when they actually produce the same index. For example, `{foo: 1}` and `{foo: 2}` actually produce the same index, but migration-verifier will flag them as mismatches. Eventually we hope to fix that, but for now false positives are safer than incomplete checks.

This also replaces use of mongo.CollectionSpecification with a struct that allows detection of unrecognized fields. Thus we’ll avoid a similar problem for collection specifications. 

This also copies over mongosync’s Option implementation.

### CODE REVIEW

File: option/bson.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	if !exists {
+		return bson.MarshalValue(primitive.Null{})
+	}
+
+	return bson.MarshalValue(val)
+}
+
+// UnmarshalBSONValue implements bson.ValueUnmarshaler.
+func (o *Option[T]) UnmarshalBSONValue(bType bsontype.Type, raw []byte) error {
+
```
Comment (tdq45gj) [inline option/bson.go:22]: nit:
```suggestion
```

File: internal/verifier/migration_verifier.go [hunk 2]
```diff
@@ -875,64 +866,74 @@ func (verifier *Verifier) compareCollectionSpecifications(srcNs string, dstNs st
...
-
-	if !reflect.DeepEqual(srcSpec.ExpireAfterSeconds, dstSpec.ExpireAfterSeconds) {
-		results = append(results, VerificationResult{
-			NameSpace: dstSpec.Namespace,
-			Cluster:   ClusterTarget,
-			ID:        dstSpec.Name,
-			Field:     "ExpireAfterSeconds",
-			Details:   Mismatch + fmt.Sprintf(" : src: %s, dst: %s", nilableToString(srcSpec.ExpireAfterSeconds), nilableToString(dstSpec.ExpireAfterSeconds))})
+	// Next check to see if the only differences are type differences.
+	// If we didn’t support pre-v5 servers we could use a $documents aggregation,
```
Comment (tdq45gj) [inline internal/verifier/migration_verifier.go:876]: I think we can reasonably limit the metadata cluster to post-v5 versions, but probably doesn't worth to do only for `$document`.

**Review by tdq45gj — APPROVED** [review]
LGTM

File: internal/verifier/migration_verifier.go (hunk 2 above)
Comment (FGasper) [reply internal/verifier/migration_verifier.go:876]: ahhh, right! I’d overlooked that this is the metadata cluster. Totally worth it, then, IMO.

**Pull request merged**